### PR TITLE
Added IPv6 and IPv4/IPv6 support to GTPv1

### PR DESCRIPTION
**Pull request by decastromonteiro**
Hey guys, I had to close my previous pull request and open a new one because it was so far behind.
@p-l- @guedou  I'm very sorry for the huge delay on fixing everything.

So we don't lose history, below I quote @alexsult :

> Hi @decastromonteiro ,
> 
> First of all, thanks for improving the GTP support. Like @p-l- , I have few things to say :
> 
>     The IE_EndUserAddress field may be already used by other scapy scripts/programms, so changing PDPAdress to PDPAddress_IPv4 (even if it is more appropriate) would break existing code. Could you keep the name PDPAdress for IPv4 and use PDPAddress_IPv6 for IPv6 ?
>     To continuously check if commits don't affect old functions, it is asked to write tests corresponding to (all) the changes you do.
>     Here is how you can check your code is covered (with the coverage module):
>     To update coverage :
> 
>     PYTHONPATH=. coverage run scapy/tools/UTscapy.py -P "load_contrib('gtp')" -t scapy/contrib/gtp.uts
> 
> And to check:
> 
>     $ coverage report scapy/contrib/gtp.py
>     Name Stmts Miss Cover
>     scapy/contrib/gtp.py 301 41 86%
> 
> Before your modifications, it was 87% covered. We don't ask you to add tests for everything, but could you add them for what you did ?
> 
> To see what is covered:
> 
>     $ coverage html scapy/contrib/gtp.py
> 
> It produces an html report in the htmlcov of your current directory.
> 
> About .anwsers and .hashret :
> 
>     Hashret
> 
>     hashret(self)
>     Must returns a string that has the same value for a request and its answer.The result of this function is used as a hash value to speed up the look fora the request matching a given response.
> 
> You are using .seq with does not exist in GTPDeletePDPContextResponse (so it will not work), but this field exists in GTPHeader and I would use it (seq from the GTPHeader) with the sgsn and ggsn IP addresses to construct this hash value.
> 
>     Answers
> 
>     answers(self,other)
>     True if self is an answer from other
> 
> I think that you can compare the results of the .hashret of the two packets
> 
>     I have seen you removed trailing spaces, that's nice regarding the pep8, but it makes difficult to read the diffs and to keep track of the modifications, could you cancel this changes ?
> 
> Alexis

1 - Alexis, I followed your suggestion and kept PDPAddress field, only adding IPv6_PDPAddress Field to support both IPv6 and IPv4/IPv6.

2 - 
> You are using .seq with does not exist in GTPDeletePDPContextResponse (so it will not work), but this field exists in GTPHeader and I would use it (seq from the GTPHeader) with the sgsn and ggsn IP addresses to construct this hash value.

It's true that sequence only exists on GTPHeader, but as GTPCreatePDPContextRequest/Response uses sequence to hashret and answers, and works fine by my tests (I'm using a real GGSN to send those messages to monitor if they're are "healthy") I also added the same methods to GTPDeletePDPContextRequest/Response. It worked fine within my scenario. Before those lines, scapy would not recognize the answer.

When trying to generate a test on scapy to verify hashret and answers, I get the error saying that seq does not exists, as you said it would. So to avoid further trouble I'm pushing this pull request without those changes.

Would you be willing to collaborate with me to solve it? I can provide real traces from GGSN and SGSN.
Let me know if you're interested.

 3 - About coverage, it's maintained on 88% after changes

Name                   Stmts   Miss  Cover
scapy/contrib/gtp.py     292     36    88%

Best regards

### CODE REVIEW

**Comment by alexsult** [comment]
On 02/23/2018 01:59 AM, Leonardo Monteiro wrote:
>
> Hey guys,
>
Hi @decastromonteiro <https://github.com/decastromonteiro> ,


> 1 - Alexis, I followed your suggestion and kept PDPAddress field, only
> adding IPv6_PDPAddress Field to support both IPv6 and IPv4/IPv6.
>
Nice, thanks.

> 2 -When trying to generate a test on scapy to verify hashret and
> answers, I get the error saying that seq does not exists, as you said
> it would. So to avoid further trouble I'm pushing this pull request
> without those changes.
>
That's nice like that. Actually, I answered too quickly last time
(sorry), this elements don't need to implement hashret and answers.

> Would you be willing to collaborate with me to solve it? I can provide
> real traces from GGSN and SGSN.
> Let me know if you're interested.
>
We can work on that. If you want, we can improve the hashret and answer
of the GTPHeader class (it should return something based on the TEID, I
think).

> 3 - About coverage, it's maintained on 88% after changes
>
> Name Stmts Miss Cover
> scapy/contrib/gtp.py 292 36 88%
>
Great, lets close this PR

Best,

Alexis

**Comment by decastromonteiro** [comment]
Hi @alexsult !

> We can work on that. If you want, we can improve the hashret and answer of the GTPHeader class (it should return something based on the TEID, I think).

So, I think that too, that's what I've come up with so far, on GTPHeader class I've defined answers, like below:

`
       def answers(self, other):
 
       if self.payload.gtp_type == 16:

          if self.seq:
            return (isinstance(other, GTPHeader) and
                    self.version == other.version and
                    self.seq == other.seq and
                    self.payload.answers() == other.teid and
                    self.payload.answer_gtp_type == other.payload.gtp_type)

        if self.seq:
          return (isinstance(other, GTPHeader) and
                  self.version == other.version and
                  self.payload.answers(other.payload) and
                  self.seq == other.seq)

        return (isinstance(other, GTPHeader) and
                self.version == other.version and
                self.payload.answers(other.payload)
                )`

So basically, when we have a GTPCreatePDPContextRequest the answer will try to match the TEICP of the Request with the TEIDI of the Response. Below is the answers function inside CreatePDPCtx class.

`    def answers(self):
      """Returns the TEID Control Plane to be compared with GTPHeader's TEID"""

      for IE in self.IE_list:

        if IE.ietype == 17:

          return IE.TEICI`

Let me know how we could work on this... Should I open an issue? another PR? Or shall we have this discussion in another forum?

BR,

Leonardo

**Comment by alexsult** [comment]
You're right, the TEID is not defined in the "create pdp request".
In the TS 29.060 / Reliable Delivery of Signalling Messages, we find:

> Each path maintains a queue with  signalling messages to be sent to the peer. The message at the front of the queue, if it is a request for which a response has been defined, shall be sent with a Sequence Number, and shall be held in a path list until a response is received. Each path has its own list. The Sequence Number shall be unique for each outstanding request message sourced from the same IP/UDP endpoint. A GSN or RNC may have several outstanding requests while waiting for responses.

1. Hashret  

Therefore, we can start with defining the hashret of GTPHeader like that (and add the version number to be "sure":

```
def hashret(self):
  return struct.pack("B", self.version) +\
    struct.pack("I", self.seq)
```
What do you think ?

We can remove the hashrets and answers from the other classes (like GTPUpdatePDPContextResponse) (because they are useless).

2. Answers
For the answers, I would make it simple, and only keep the following part of your code:

```
     return (isinstance(other, GTPHeader) and
              self.version == other.version and
              self.seq == other.seq

```

Since there is no answers implemented for the upper layers, I think we can remove `self.payload.answers(other.payload)`

Tell me what you think about that, feel free to open an issue if you want to discuss about it.

Regards,

Alexis
